### PR TITLE
Rely on the new case representation in Evarsolve.

### PR DESCRIPTION
**Pull request by ppedrot**

### CODE REVIEW

**Comment by ppedrot** [comment]
Bench: https://gitlab.com/coq/coq/-/jobs/1941587259

**Comment by ppedrot** [comment]
```
┌─────────────────────────────┬─────────────────────────┬───────────────────────────────────────┬───────────────────────────────────────┬─────────────────────────┬───────────────────┐
│                             │      user time [s]      │              CPU cycles               │           CPU instructions            │  max resident mem [KB]  │    mem faults     │
│                             │                         │                                       │                                       │                         │                   │
│                package_name │     NEW      OLD  PDIFF │            NEW             OLD  PDIFF │            NEW             OLD  PDIFF │     NEW      OLD  PDIFF │ NEW  OLD    PDIFF │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼───────────────────┤
│               coq-fourcolor │ 1491.85  1507.92  -1.07 │  6598084588186   6663326876532  -0.98 │ 12088567051221  12088798168645  -0.00 │  726388   726556  -0.02 │   0    0      nan │
│          coq-mathcomp-field │  113.47   114.09  -0.54 │   518411953727    521138830712  -0.52 │   845542071446    853576692309  -0.94 │  649508   658748  -1.40 │   0    0      nan │
│              coq-verdi-raft │  579.12   582.24  -0.54 │  2637055929045   2649996464347  -0.49 │  4048104041629   4047952226235   0.00 │ 1220752  1222904  -0.18 │   0    0      nan │
│               coq-fiat-core │   54.24    54.34  -0.18 │   235684557356    233402130553   0.98 │   323869783875    323978453499  -0.03 │  473816   473620   0.04 │   0    0      nan │
│       coq-mathcomp-fingroup │   23.34    23.38  -0.17 │   106413667752    106154885344   0.24 │   155468801945    155429273094   0.03 │  479996   477864   0.45 │   0    0      nan │
│                 coq-coqutil │   34.77    34.77   0.00 │   155340974020    155537807891  -0.13 │   207263613597    207209304834   0.03 │  601140   602016  -0.15 │   0    0      nan │
│               coq-perennial │ 4148.41  4146.96   0.03 │ 18822571447759  18815076489642   0.04 │ 30822874183513  30817192502762   0.02 │ 2612724  2612812  -0.00 │   0    4  -100.00 │
│      coq-mathcomp-character │   76.78    76.74   0.05 │   350445923691    350081701481   0.10 │   528731988468    528709425727   0.00 │  724152   726372  -0.31 │   0    0      nan │
│                coq-rewriter │  375.67   375.20   0.13 │  1701067978187   1699809827361   0.07 │  2771254088344   2771174538804   0.00 │ 1060032  1062980  -0.28 │   0    0      nan │
│                   coq-color │  265.37   264.82   0.21 │  1193028312711   1192344968211   0.06 │  1599219365721   1599349010047  -0.01 │ 1138516  1138148   0.03 │   0    0      nan │
│                 coq-bignums │   28.33    28.27   0.21 │   128513074095    128736546751  -0.17 │   175932008749    175946385725  -0.01 │  468856   466492   0.51 │   0    0      nan │
│ coq-rewriter-perf-SuperFast │  927.55   925.21   0.25 │  4177622151318   4168157830497   0.23 │  7195458736446   7194717244504   0.01 │ 1144792  1115412   2.63 │   0    0      nan │
│                   coq-verdi │   48.78    48.65   0.27 │   220277122543    219608493820   0.30 │   324448628711    324385835314   0.02 │  827188   828328  -0.14 │   0    0      nan │
│                    coq-corn │  802.48   800.20   0.28 │  3643421904290   3630495275375   0.36 │  5581221966688   5589597658242  -0.15 │  878748   875380   0.38 │   0    0      nan │
│             coq-fiat-crypto │ 2949.31  2940.07   0.31 │ 13305702451259  13262675201961   0.32 │ 24369571456664  24368052555213   0.01 │ 3855104  3855136  -0.00 │   0    0      nan │
│                  coq-geocoq │  717.48   715.05   0.34 │  3261372907271   3250047776499   0.35 │  5155671274598   5155984122646  -0.01 │ 1074592  1076120  -0.14 │   0    0      nan │
│            coq-fiat-parsers │  331.37   329.98   0.42 │  1472017048287   1465281895860   0.46 │  2422101623635   2421918067212   0.01 │ 2790112  2789440   0.02 │   0    0      nan │
│                coq-compcert │  292.23   290.94   0.44 │  1314832119141   1310246086536   0.35 │  1943741587523   1943592910008   0.01 │ 1097084  1098808  -0.16 │   0    0      nan │
│       coq-engine-bench-lite │  177.93   177.06   0.49 │   763770280540    759313011064   0.59 │  1484611523092   1482600635091   0.14 │ 1175768  1174024   0.15 │   0    0      nan │
│                  coq-stdlib │  407.25   405.22   0.50 │  1652890511532   1648655168999   0.26 │  1445340724193   1445421338850  -0.01 │  590828   588868   0.33 │   0    0      nan │
│           coq-iris-examples │  466.08   463.69   0.52 │  2113886680276   2101763106422   0.58 │  3142898594801   3142914299150  -0.00 │ 1181232  1182916  -0.14 │   0    0      nan │
│  coq-performance-tests-lite │  788.45   783.96   0.57 │  3542797643957   3520974160481   0.62 │  5983991560700   5983474036009   0.01 │ 1826168  1825896   0.01 │   3    0      nan │
│              coq-coquelicot │   34.83    34.63   0.58 │   155352517188    154256671878   0.71 │   202569939831    202611471624  -0.02 │  746912   749384  -0.33 │   0    0      nan │
│                    coq-core │   97.24    96.66   0.60 │   388219745499    383531570263   1.22 │   432975779426    432995360608  -0.00 │  249036   249788  -0.30 │   0    0      nan │
│                  coq-flocq3 │   76.98    76.49   0.64 │   348645876258    346555200194   0.60 │   454785517433    454887078909  -0.02 │  988184   991560  -0.34 │   0    0      nan │
│                 coq-unimath │ 3434.37  3411.99   0.66 │ 15625899565989  15520483143233   0.68 │ 30937304419058  30938969146012  -0.01 │ 3859688  3858772   0.02 │   0    0      nan │
│        coq-mathcomp-algebra │   63.94    63.52   0.66 │   291141707446    289078173520   0.71 │   398436699203    399701492632  -0.32 │  558576   558560   0.00 │   0    0      nan │
│                    coq-hott │  154.05   152.95   0.72 │   694059053873    688374955658   0.83 │  1076351200628   1076292812848   0.01 │  658588   654836   0.57 │   0    0      nan │
│                     coq-vst │ 1155.82  1146.05   0.85 │  5243900121826   5199139623698   0.86 │  8415849647361   8416005428826  -0.00 │ 1898300  1897704   0.03 │  13    0      nan │
│            coq-math-classes │   94.39    93.59   0.85 │   425387800041    421063587758   1.03 │   592562919309    592558393230   0.00 │  499628   495648   0.80 │   0    0      nan │
│      coq-mathcomp-odd-order │  564.78   559.47   0.95 │  2583368892296   2558922709622   0.96 │  4328730931706   4328542249199   0.00 │  937904   933600   0.46 │   0    0      nan │
│                coq-bedrock2 │  190.45   188.56   1.00 │   854096256050    846018973137   0.95 │  1567468454103   1568151566962  -0.04 │ 2053896  2054132  -0.01 │   1    0      nan │
│      coq-mathcomp-ssreflect │   26.09    25.79   1.16 │   118214024157    116473071848   1.49 │   147707959451    147740242931  -0.02 │  533656   531428   0.42 │   0    0      nan │
│       coq-mathcomp-solvable │   88.61    87.29   1.51 │   404062330582    398034983297   1.51 │   613481973062    613439634506   0.01 │  670808   670716   0.01 │   0    0      nan │
│                coq-coqprime │  182.70   177.76   2.78 │   833273709658    810735903179   2.78 │  1559822510997   1559714628413   0.01 │  854260   854972  -0.08 │   0    0      nan │
└─────────────────────────────┴─────────────────────────┴───────────────────────────────────────┴───────────────────────────────────────┴─────────────────────────┴───────────────────┘
```

**Comment by ppedrot** [comment]
Apart from the following line that indicates quite a speedup in mathcomp-field, I don't see anything suspicious in the per-diff logs.
```
html/coq-mathcomp-field/mathcomp/field/algebraics_fundamentals.v.html:193 -1.07 1.26 0.18 (-86%)
```
I'd be inclined to blame the small slowdown in the two mathcomp devs onto noise, but just to be sure I'll rerun a bench.

**Comment by ppedrot** [comment]
Another try at the same bench: https://gitlab.com/coq/coq/-/jobs/1943768584

**Comment by ppedrot** [comment]
```
┌─────────────────────────────┬─────────────────────────┬───────────────────────────────────────┬───────────────────────────────────────┬─────────────────────────┬───────────────────┐
│                             │      user time [s]      │              CPU cycles               │           CPU instructions            │  max resident mem [KB]  │    mem faults     │
│                             │                         │                                       │                                       │                         │                   │
│                package_name │     NEW      OLD  PDIFF │            NEW             OLD  PDIFF │            NEW             OLD  PDIFF │     NEW      OLD  PDIFF │ NEW  OLD    PDIFF │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼───────────────────┤
│               coq-fourcolor │ 1510.37  1522.16  -0.77 │  6606594349119   6658672850614  -0.78 │ 12086756554864  12086874418303  -0.00 │  724412   724448  -0.00 │   0    0      nan │
│          coq-mathcomp-field │  116.89   117.22  -0.28 │   512165111592    513772787496  -0.31 │   843519444032    851153722537  -0.90 │  657740   655144   0.40 │   0    0      nan │
│        coq-mathcomp-algebra │   65.79    65.87  -0.12 │   287490344897    288044051399  -0.19 │   397820272305    399038496911  -0.31 │  560212   559544   0.12 │   0    0      nan │
│                 coq-coqutil │   35.95    35.93   0.06 │   155367441492    154292288469   0.70 │   207293470454    207298366320  -0.00 │  601100   601256  -0.03 │   4    0      nan │
│ coq-rewriter-perf-SuperFast │  953.70   952.89   0.09 │  4168792445687   4165381471473   0.08 │  7193700804084   7193052563514   0.01 │ 1145016  1146508  -0.13 │  26    0      nan │
│  coq-performance-tests-lite │  805.58   804.43   0.14 │  3505238922538   3501009221537   0.12 │  5958987833996   5959562617035  -0.01 │ 1826080  1825820   0.01 │   1    0      nan │
│                 coq-unimath │ 3538.05  3531.94   0.17 │ 15492010147686  15464596580939   0.18 │ 30944101505250  30941115242063   0.01 │ 3801148  3801244  -0.00 │   2    5   -60.00 │
│                  coq-stdlib │  420.63   419.75   0.21 │  1515769349074   1524596487759  -0.58 │  1445641747453   1445807410373  -0.01 │  590240   589920   0.05 │   0    0      nan │
│      coq-mathcomp-odd-order │  580.71   579.45   0.22 │  2545787251593   2540507271799   0.21 │  4309202994710   4309163530326   0.00 │  934400   932896   0.16 │   0    0      nan │
│                coq-bedrock2 │  195.03   194.49   0.28 │   851234514172    849496384415   0.20 │  1567994968368   1567906684591   0.01 │ 2053888  2053824   0.00 │  20   18    11.11 │
│                coq-rewriter │  388.73   387.58   0.30 │  1696142266741   1691028683553   0.30 │  2770110118678   2770795735464  -0.02 │ 1060180  1063312  -0.29 │  14   11    27.27 │
│                coq-compcert │  302.97   301.82   0.38 │  1316632968580   1312752992970   0.30 │  1944055386010   1944148548707  -0.00 │ 1097292  1097092   0.02 │   0    8  -100.00 │
│              coq-verdi-raft │  603.81   601.21   0.43 │  2638693057868   2625161049651   0.52 │  4047682666828   4047454508768   0.01 │ 1217732  1218196  -0.04 │   0    0      nan │
│            coq-fiat-parsers │  341.25   339.48   0.52 │  1475109809236   1467954137788   0.49 │  2422001745955   2421581539305   0.02 │ 2787628  2787340   0.01 │   0    1  -100.00 │
│            coq-math-classes │   97.35    96.84   0.53 │   422710708475    420715841132   0.47 │   591560723206    591549698701   0.00 │  495476   495592  -0.02 │   0    0      nan │
│                    coq-corn │  826.04   821.68   0.53 │  3608119374766   3586885593534   0.59 │  5579707698670   5587690756615  -0.14 │  876836   876236   0.07 │   0    0      nan │
│                    coq-hott │  158.65   157.76   0.56 │   688948943327    683461464795   0.80 │  1075467736688   1075377703822   0.01 │  659156   658108   0.16 │   0    0      nan │
│                     coq-vst │ 1193.02  1186.16   0.58 │  5213800882811   5182493470984   0.60 │  8416939183117   8417317008466  -0.00 │ 1894732  1896264  -0.08 │   2    0      nan │
│                  coq-flocq3 │   78.98    78.51   0.60 │   344080949020    342271246360   0.53 │   455182385764    455167319821   0.00 │  987864   990460  -0.26 │   8    0      nan │
│             coq-fiat-crypto │ 3062.42  3043.71   0.61 │ 13364699669769  13283834575931   0.61 │ 24364594129336  24365093391946  -0.00 │ 3852688  3852616   0.00 │  72  109   -33.94 │
│      coq-mathcomp-ssreflect │   26.84    26.67   0.64 │   116681492276    116505723856   0.15 │   147428768341    147509270112  -0.05 │  531276   535624  -0.81 │   0    0      nan │
│                  coq-geocoq │  746.33   741.59   0.64 │  3258289273613   3240261144734   0.56 │  5156732973725   5156802481108  -0.00 │ 1075924  1078412  -0.23 │   2    0      nan │
│           coq-iris-examples │  478.84   475.63   0.67 │  2090522205485   2077783032875   0.61 │  3139756093350   3139568534424   0.01 │ 1180856  1180872  -0.00 │   2    2     0.00 │
│       coq-mathcomp-solvable │   91.09    90.44   0.72 │   398307424923    395537730815   0.70 │   612202156827    612104651841   0.02 │  668924   670240  -0.20 │   0    0      nan │
│                coq-coqprime │  188.65   187.27   0.74 │   825628750576    819884713395   0.70 │  1559900935662   1559921516145  -0.00 │  857592   859620  -0.24 │   7    0      nan │
│                   coq-verdi │   50.36    49.97   0.78 │   218634325906    216486334522   0.99 │   324129737190    324124295885   0.00 │  827272   828612  -0.16 │   0    0      nan │
│      coq-mathcomp-character │   79.52    78.87   0.82 │   348423748746    345429257150   0.87 │   527791927459    527762621668   0.01 │  732116   734252  -0.29 │   0    0      nan │
│               coq-perennial │ 4279.46  4242.97   0.86 │ 18716025252797  18554969257214   0.87 │ 30681059890233  30682666212006  -0.01 │ 2599836  2601336  -0.06 │   0    0      nan │
│              coq-coquelicot │   35.88    35.56   0.90 │   154779013981    152911550295   1.22 │   202585959008    202600899046  -0.01 │  746868   749492  -0.35 │ 147    0      nan │
│                   coq-color │  274.40   271.83   0.95 │  1190568715604   1180400085734   0.86 │  1599970956148   1599903020234   0.00 │ 1139164  1137464   0.15 │   2    0      nan │
│       coq-mathcomp-fingroup │   24.39    24.16   0.95 │   106512292150    105520657414   0.94 │   155230502031    155225376685   0.00 │  480180   477948   0.47 │   0    0      nan │
│                    coq-core │  101.86   100.87   0.98 │   375936250045    372301934894   0.98 │   432951799680    433088817661  -0.03 │  249896   248720   0.47 │   0    0      nan │
│                 coq-bignums │   29.63    29.34   0.99 │   128870560175    128001468301   0.68 │   176061829341    176068619083  -0.00 │  465420   465936  -0.11 │   0    0      nan │
│       coq-engine-bench-lite │  182.18   180.01   1.21 │   765942022439    755318909885   1.41 │  1485338770276   1485596508457  -0.02 │ 1173944  1173844   0.01 │   0    0      nan │
│               coq-fiat-core │   56.05    55.07   1.78 │   235025908089    231081291361   1.71 │   324264682286    324333827753  -0.02 │  473856   473800   0.01 │   6    6     0.00 │
└─────────────────────────────┴─────────────────────────┴───────────────────────────────────────┴───────────────────────────────────────┴─────────────────────────┴───────────────────┘
```

**Comment by ppedrot** [comment]
It does look like the bench infrastructure with the new machines is still pretty noisy...

**Comment by ppedrot** [comment]
FTR, the mathcomp speedup in a single line is still there, and there is nothing else in sight in per-line diffs, again. I consider this ready then.

File: pretyping/evarconv.ml [hunk 1]
```diff
@@ -380,12 +441,18 @@ let rec ise_stack2 no_app env evd f sk1 sk2 =
...
+      let (ci1, u1, pms1, t1, br1) = Stack.expand_case env evd cse1 in
+      let (ci2, u2, pms2, t2, br2) = Stack.expand_case env evd cse2 in
+      let hd1 = mkIndU (ci1.ci_ind, u1) in
+      let hd2 = mkIndU (ci2.ci_ind, u2) in
+      let fctx i (ctx1, t1) (_ctx2, t2) = f (push_rel_context ctx1 env) i CONV t1 t2 in
       begin
         match ise_and i [
-          (fun i -> f env i CONV t1 t2);
-          (fun i -> ise_array2 i (fun ii -> f env ii CONV) c1 c2)]
+          (fun i -> compare_heads env i ~nargs:None hd1 hd2);
```
Comment (SkySkimmer) [inline pretyping/evarconv.ml:451]: Is `None` correct here?

Comment (ppedrot) [reply pretyping/evarconv.ml:451]: Yes, we know that the inductive is always fully applied in the return clause.

Comment (SkySkimmer) [reply pretyping/evarconv.ml:451]: oh None means "fully applied"
That's kinda weird
How about having `type nargs = FullyApplied | SomeArgs of int`?

Comment (ppedrot) [reply pretyping/evarconv.ml:451]: This API is not exported, I don't think it's worthwhile to introduce a new type for this.

Comment (SkySkimmer) [reply pretyping/evarconv.ml:451]: The current usage of option is misleading IMO, please make a new type.

File: pretyping/evarconv.ml [hunk 2]
```diff
@@ -358,6 +358,67 @@ let rec ise_app_rev_stack2 env f evd revsk1 revsk2 =
...
+let compare_cumulative_instances evd variances u u' =
+  match Evarutil.compare_cumulative_instances CONV variances u u' evd with
+  | Inl evd ->
+    Success evd
+  | Inr p -> UnifFailure (evd, UnifUnivInconsistency p)
+
+let is_int_opt o n = match o with None -> true | Some m -> Int.equal m n
+
+let compare_heads env evd ~nargs term term' =
+    let check_strict evd u u' =
```
Comment (SkySkimmer) [inline pretyping/evarconv.ml:372]: While you're moving this code please reindent the whole function.

**Comment by ppedrot** [comment]
@SkySkimmer nitpicks addressed.

**Comment by ppedrot** [comment]
@SkySkimmer would you assign?

**Review by SkySkimmer — APPROVED** [review]

**Comment by SkySkimmer** [comment]
@coqbot merge now

**Pull request merged**